crates/biome_cli/src/reporter/gitlab.rs

```diff
@@ -0,0 +1,243 @@
+                let mut content = serde_json::to_string(&diagnostic)?;
+                if index < total_diagnostics - 1 {
+                    content.push(',')
+                }
+
+                self.console.log(markup!({ content }));
+            }
+        }
+        self.console.log(markup!("]"));
+        Ok(())
+    }
+}
+
+pub(crate) struct GitLabDiagnosticBuilder {
+    /// The root of the Git repository.
+    /// Required for relativization of the reported absolute paths.
+    repository_root: Option<PathBuf>,
+    /// A set of fingerprints (unique identifiers) to prevent collisions.
+    fingerprints: HashSet<u64>,
+}
+
+impl GitLabDiagnosticBuilder {
+    pub fn new(repository_root: Option<PathBuf>) -> Self {
+        Self {
+            repository_root,
+            fingerprints: HashSet::new(),
+        }
+    }
+
+    /// Turns a biome diagnostic into one in the GitLab Code Quality report format.
+    pub fn gitlab_diagnostic(
+        &mut self,
+        value: biome_diagnostics::error::Error,
+    ) -> GitLabDiagnostic {
+        let check_name = self.check_name(&value).unwrap_or_default();
+        let path = self.path(&value).unwrap_or_default();
+        let line = self.line(&value).map_or(1, |line| max(line, 1));
+        let source_code = self.source_code(&value).unwrap_or_default();
+        let fingerprint = self
+            .ensure_fingerprint_uniqueness(self.fingerprint(&check_name, &path, &source_code), 0)
+            .to_string();
+
+        GitLabDiagnostic {
+            severity: GitLabSeverity::from(value.severity()),
+            description: PrintDescription(&value).to_string(),
+            check_name,
+            fingerprint,
+            location: Location {
+                path,
+                lines: Lines { begin: line },
+            },
+        }
+    }
+
+    /// Extracts the name of the category as the check name.
+    fn check_name(&self, value: &biome_diagnostics::error::Error) -> Option<String> {
+        let category = value.category()?;
+        Some(category.name().to_string())
+    }
+
+    /// Extracts the source code generating the diagnostic.
+    fn source_code(&self, value: &biome_diagnostics::error::Error) -> Option<String> {
+        let location = value.location();
+        let source_code = location.source_code?;
+        let span = location.span?;
+
+        Some(source_code.text[span].to_string())
+    }
+
+    /// Generates a fingerprint for a diagnostic.
+    fn fingerprint(&self, check_name: &String, path: &String, code: &String) -> u64 {
+        let mut hasher = DefaultHasher::new();
+
+        // Including the source code in our hash leads to more stable
+        // fingerprints. If you instead rely on e.g. the line number and change
+        // the first line of a file, all of its fingerprint would change.
+        code.hash(&mut hasher);
+        check_name.hash(&mut hasher);
+        path.hash(&mut hasher);
+
+        hasher.finish()
```

Comment:
A more idiomatic way to calculate the hash is to create a type called `Fingerprint` that accepts these three arguments and then add `#[derive(Hash)]`.

Then you can have a method that does this

```rs
fn calculate_hash<T: Hash>(t: &T) -> u64 {
    let mut s = DefaultHasher::new();
    t.hash(&mut s);
    s.finish()
}
```